core/src/main/java/io/grpc/internal/AbstractServerStream.java

```diff
@@ -31,256 +31,215 @@
+      this.listener = Preconditions.checkNotNull(listener);
 
-  /**
-   * Indicates the stream is considered completely closed and there is no further opportunity for
-   * error. It calls the listener's {@code closed()} if it was not already done by {@link
-   * #abortStream}. Note that it is expected that either {@code closed()} or {@code abortStream()}
-   * was previously called, since {@code closed()} is required for a normal stream closure and
-   * {@code abortStream()} for abnormal.
-   */
-  public void complete() {
-    if (!gracefulClose) {
-      closeListener(Status.INTERNAL.withDescription("successful complete() without close()"));
-      throw new IllegalStateException("successful complete() without close()");
+      // Now that the stream has actually been initialized, call the listener's onReady callback if
+      // appropriate.
+      onStreamAllocated();
     }
-    closeListener(Status.OK);
-  }
 
-  /**
-   * Called when the remote end half-closes the stream.
-   */
-  @Override
-  protected final void remoteEndClosed() {
-    halfCloseListener();
-  }
+    @Override
+    public void deliveryStalled() {}
 
-  /**
-   * Aborts the stream with an error status, cleans up resources and notifies the listener if
-   * necessary.
-   *
-   * <p>Unlike {@link #close(Status, Metadata)}, this method is only called from the
-   * transport. The transport should use this method instead of {@code close(Status)} for internal
-   * errors to prevent exposing unexpected states and exceptions to the application.
-   *
-   * @param status the error status. Must not be {@link Status#OK}.
-   * @param notifyClient {@code true} if the stream is still writable and you want to notify the
-   *        client about stream closure and send the status
-   */
-  public final void abortStream(Status status, boolean notifyClient) {
-    // TODO(louiscryan): Investigate whether we can remove the notification to the client
-    // and rely on a transport layer stream reset instead.
-    Preconditions.checkArgument(!status.isOk(), "status must not be OK");
-    closeListener(status);
-    if (notifyClient) {
-      // TODO(louiscryan): Remove
-      if (stashedTrailers == null) {
-        stashedTrailers = new Metadata();
-      }
-      writeStatusToTrailers(status);
-      sendStreamAbortToClient(status, stashedTrailers);
+    @Override
+    public void endOfStream() {
+      closeDeframer();
+      listener().halfClosed();
     }
-  }
 
-  @Override
-  public boolean isClosed() {
-    return super.isClosed() || listenerClosed;
-  }
+    @Override
+    protected ServerStreamListener listener() {
+      return listener;
+    }
 
-  /**
-   * Notifies the remote client that this stream has aborted.
-   */
-  protected abstract void sendStreamAbortToClient(Status status, Metadata trailers);
+    /**
+     * Called in the network thread to process the content of an inbound DATA frame from the client.
```

Comment:
I know you didn't write this, but should we change it to "transport thread", to have the same terminology everywhere?